Grammar/python.gram

```diff
@@ -1345,6 +1347,14 @@ invalid_except_stmt:
     | a='except' '*'? expression ['as' NAME ] NEWLINE { RAISE_SYNTAX_ERROR("expected ':'") }
     | a='except' NEWLINE { RAISE_SYNTAX_ERROR("expected ':'") }
     | a='except' '*' (NEWLINE | ':') { RAISE_SYNTAX_ERROR("expected one or more exception types") }
+invalid_except_as_expr:
+    | 'except' expression 'as' a=expression {
+        RAISE_SYNTAX_ERROR_KNOWN_LOCATION(
+            a, "cannot use except statement with %s", _PyPegen_get_expr_name(a)) }
+invalid_except_star_as_expr:
+    | 'except' '*' expression 'as' a=expression {
+        RAISE_SYNTAX_ERROR_KNOWN_LOCATION(
+            a, "cannot use except* statement with %s", _PyPegen_get_expr_name(a)) }
```

Comment:
These two could be additional alternatives in the `invalid_except_stmt` rule above. They just have to appear first.

Reply:
Sorry, I don't understand your comment :(
Can you please specify?

Reply:
Sorry for not making myself clear. The idea was that we don't need extra rules (`invalid_except_as_expr` and `invalid_except_star_as_expr`), we can just add two more alternatives to the already existing rule `invalid_except_stmt`. Does that make sense?

Reply:
@lysnikolaou yes, this was my initial idea. As I said in https://github.com/python/cpython/pull/123442#issue-2492869658

But, if I do that this happens:

```
Failed example:
    try:
      pass
    except* A as a:
      pass
    except*:
      pass
Expected:
    Traceback (most recent call last):
    SyntaxError: expected one or more exception types
Got:
    Traceback (most recent call last):
      File "/Users/sobolev/Desktop/cpython2/Lib/doctest.py", line 1395, in __run
        exec(compile(example.source, filename, "single",
             ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                     compileflags, True), test.globs)
                     ^^^^^^^^^^^^^^^^^^^
      File "<doctest test.test_syntax[305]>", line 3
        except* A as a:
                     ^
    SyntaxError: cannot use except statement with name


test_syntax failed (1 failure)
```

This is why this happens:
1. `invalid_except_stmt` is called in both `except_block` and `except_star_block`
2. They are called in

```
try_stmt[stmt_ty]:
    | invalid_try_stmt
    | 'try' &&':' b=block f=finally_block { _PyAST_Try(b, NULL, NULL, f, EXTRA) }
    | 'try' &&':' b=block ex[asdl_excepthandler_seq*]=except_block+ el=[else_block] f=[finally_block] { _PyAST_Try(b, ex, el, f, EXTRA) }
    | 'try' &&':' b=block ex[asdl_excepthandler_seq*]=except_star_block+ el=[else_block] f=[finally_block] {
        CHECK_VERSION(stmt_ty, 11, "Exception groups are",
                      _PyAST_TryStar(b, ex, el, f, EXTRA)) }
```

3. Here `except_block` is called first, on the second run of the parser, `invalid_` rules are also executed and that's why the first match is our new `'except' expression 'as' a=expression` rule which raises this `SyntaxError`

Does this seem correct to you?

Reply:
Hmm, that's a good catch. It sounds like this means we should have two different rules for invalid `except` and `except*`. So have an `invalid_except_stmt` that is called from `except_block`, where all the `'*'?` have been stripped, and then an `invalid_except_star_stmt` that is called from `except_star_block` and the `'*'` is no longer optional.

What do you think?

Reply:
Great idea, refactored! 👍